### PR TITLE
Navigation features and lots more stuff

### LINKED ISSUE #37: Detect odometry problems
move_basic detect if a robot is not reporting odometry consistent with the way it is being asked to move and abort with an error message.

DOD: test on a robot with disconnected/cross-connected encoders and verify it detects the odometry problem.

### LINKED ISSUE #47: Create better test for navigation being complete
Definition of done: definition of done :slightly_smiling_face:

### LINKED ISSUE #74: max_lateral_rotation and max_angular_velocity parameters not clear and undocumented
Found out that parameter `max_angular_velocity` is meant for maximum angular velocity, while robot is **turning on the spot**, while the parameter `max_lateral_rotation` controls the maximum angular velocity while **driving towards goal**. This answers this issue but maybe the names of the parameters are a bit unintuitive? The [ROS wiki page](http://wiki.ros.org/move_basic) also doesn't even mention param `max_lateral_rotation`

**suggestion1:** new names: max_turning_velocity, max_lateral_velocity
**suggestion2:** we set the max_lateral_velocity = max_turning_velocity as default with 
`nh.param<double>("max_lateral_velocity", maxLateralVelocity, maxTurningVelocity);`
as this would still lave the user the control over both, but default parameters would be a bit easier to understand.

If we don't rename parameters, we should at least update documentation on ROS wiki page (we have to do that anyway).

### LINKED ISSUE #75: Add dynamic reconfigure to move basic
It would be beneficial to add params of move basic to dynamic reconfigure as it would enable faster debugging and optimising

### LINKED ISSUE #81: Robot driving past goals into infinity
In a nutshell if the robot comes into such a case where it misses a goal it doesn't stop but attempts to drive ahead and loop around to find the goal. That's obviously problematic in almost all cases.

A suggested fix would be to constantly calculate a safety dot product between the robot's forward vector and the vector from the robot center to the goal, then stop driving if the result is <= 0 and distance to goal is more than 0.3 m or something. That would prevent the robot from ever driving away from the goal. As to what to do when that condition is detected, we could just return to the initial rotation stage or cancel the goal completely and resend it.

Removing the low speed requirement for having achieved a goal is probably also a good idea, after all we're there and we need to stop right now regardless of the speed we're going at.

This is probably the highest priority issue along with #80 which is the main but not only cause of it.

### PR DESCRIPTION
**Pull request by dorkamotorka**
This add features that were also included in the breadcrumb/move_smooth. 
Besides that it fixes a bug - Robot was not turning toward the requested final orientation, because the final rotation was calculated as (initial robot orientation - final wanted robot orientation (both in odom)), but the robot arrived to the goal with different orientation as initial, therefore it didn't turned in place the way it should.

### CODE REVIEW

File: CMakeLists.txt
Comment (JanezCim) [inline CMakeLists.txt:92]: You are already doing tests for goal queueing? Awesome! Would be nice to know how to run them :))

File: cfg/Movebasic.cfg
Comment (JanezCim) [inline cfg/Movebasic.cfg]: thank you for adding (when turning on spot) its much more clear now

File: cfg/Movebasic.cfg
Comment (JanezCim) [inline cfg/Movebasic.cfg]: Maybe also add units into description - here m/s - also holds for other relevant params

File: src/move_basic.cpp
Comment (JanezCim) [inline src/move_basic.cpp]: This is totally fine, but you do know once you have dynamic reconfigure, you dont actually need nh.param anymore. The defaut params are set into ros param server from the .cfg file and then  dynamicReconfigCallback is called at initialisation to set the params in the code. Check this out: https://github.com/UbiquityRobotics/avalon/blob/master/avalon_aislemode/src/aisle_path_setter.cpp. Anyway, just recommendation for future improvements :)

File: src/move_basic.cpp
Comment (JanezCim) [inline src/move_basic.cpp:461]: Am i correct to assume there is no more rotation attempts in the new move_base?

File: src/move_basic.cpp
Comment (JanezCim) [inline src/move_basic.cpp:533]: 20Hz is fine, but is there any specific reason for it being this instead of 15?

File: src/move_basic.cpp
Comment (JanezCim) [inline src/move_basic.cpp:659]: Is planning frame not needed anymore in move linear?

File: src/move_basic.cpp
Comment (JanezCim) [inline src/move_basic.cpp:202]: Now that the names and meaning of params will be changed, its important we also update the documentation on README and ROS wiki

**Review by JanezCim — APPROVED** [review]
I've run it in Gazebo in Melodic and it works perfectly. Also tested the dynamic reconfigure - also works. My next tests will be on actual robots. This is a huge change, we need to take care that the documentation is up to date

**Comment by JanezCim** [comment]
@tp4348 where have the min_lateral_velocity and min_turning_velocity gone? is that not implemented?

File: src/move_basic.cpp
Comment (MoffKalast) [reply src/move_basic.cpp]: @JanezCim are you completely sure removing that won't break some legacy setup?

File: src/move_basic.cpp
Comment (MoffKalast) [reply src/move_basic.cpp:533]: 20 > 15, therefore it's better 😛 

For real though, someone should check cpu usage with that change, it may be significant.

**Comment by JanezCim** [comment]
and what about params like robot_front_length and robot_width, will they not be in dynamic reconfigure?

File: src/move_basic.cpp
Comment (MoffKalast) [inline src/move_basic.cpp:532]: Wouldn't this still be useful for debug?

**Review by MoffKalast — APPROVED** [review]
Tested it out in gazebo on kinetic, seems to be working well. Given that there is a change in how the system operates (goal queue instead of replacing the current goal instantly) we should document it as well as possible as to not silently break existing 3rd party software.

The change from a speed curve of /\ to /¯¯¯¯\ is also notable, but since this is far more what one would expect from the robot it's more of a fix than a change.

File: src/move_basic.cpp
Comment (JanezCim) [reply src/move_basic.cpp]: nope - that's in the "načeloma" department

File: CMakeLists.txt
Comment (dorkamotorka) [reply CMakeLists.txt:92]: @JanezCim that's the unit test you also ran on your computer. This was done 3 months back.

File: cfg/Movebasic.cfg
Comment (dorkamotorka) [reply cfg/Movebasic.cfg]: Will do, thanks

File: src/move_basic.cpp
Comment (dorkamotorka) [reply src/move_basic.cpp]: Hmm.. I do know these parameters get overridden by dynamic reconfigure. I have made all default values the same and would leave like that for convenience, if one decides to use either dynamic_reconfigure or nh.param values.

File: src/move_basic.cpp
Comment (dorkamotorka) [reply src/move_basic.cpp:461]: I have implemented a proportional controller, which should be tuned so, that this does not happen. Since I was testing move_basic I couldn't reproduce this to be actually useful, because it always stopped fine. @JanezCim If you think this feature is crucial I can re-added, no problem

File: src/move_basic.cpp
Comment (dorkamotorka) [reply src/move_basic.cpp:533]: I will check the CPU, this is actually a really important parameter, because in some cases might also affect how fast we are getting the updated distance to the object - might change the collision avoidance.

File: src/move_basic.cpp
Comment (dorkamotorka) [reply src/move_basic.cpp:659]: I did thought about this overall, because there's lots of confusion about planning frame and driving frame, because they are actually the same for our application but didn't want to delete because it might make sense for other projects. Nevertheless, in moveLinear function it didn't make sense to have the planningFrame anymore because it was used for transformation that wasn't used. But I will definitely consider this in the future, because during linear portion of the goal, we could update the pose according to some frame other than /odom or /map and get a better localization(for example fiducial)

File: src/move_basic.cpp
Comment (dorkamotorka) [reply src/move_basic.cpp:202]: Yes yes of course, I will definitely do that before the merge - actually at the same time hah

**Comment by dorkamotorka** [comment]
> @tp4348 where have the min_lateral_velocity and min_turning_velocity gone? is that not implemented?

If you have min_lateral velocity specified theoretically you cannot even drive the robot straight, because there will also be this lateral component and also for the turning velocity you cannot bring the error as close to 0 as possible. Velocities are now limited like [-max_lateral_velocity, max_lateral_velocity] and also for the turning velocity,.. This ranges are just below  what robot is able to perform in the firmware.

**Comment by dorkamotorka** [comment]
> and what about params like robot_front_length and robot_width, will they not be in dynamic reconfigure?

This values are the lengths of the robot mechanical parts. They shouldn't be dynamically reconfigured. Changing them would change the behavior of collision avoidance(stopping before collision) but for this you have forwardObstacleThreshold

**Comment by dorkamotorka** [comment]
Just to make it clear - I will make sure the READ.md and ROS Wiki is updated

File: src/move_basic.cpp
Comment (dorkamotorka) [reply src/move_basic.cpp:532]: Good point. I will add the debug stuff

File: src/move_basic.cpp
Comment (JanezCim) [reply src/move_basic.cpp:461]: No actually i dont care much about it, never needed it, just wanted to know

**Review by AGregorc — COMMENTED** [review]
I just did a quick flyby of changes since I didn't use move_basic directly yet and you already had some good reviews. So I'm just gonna say great work!  :beers:

**Comment by dorkamotorka** [comment]
@AGregorc This PR will be open at least till the end of this week. If you have time, it would be useful for you to understand it, because I will be away in January and that directly relates to navigation which we will be using for Q... .

**Comment by tp4348** [comment]
Closing this since a bigger merge will be done with `nav-features` branch.

**Comment by dorkamotorka** [comment]
Mistakenly closed it.

**Comment by tp4348** [comment]
Please re-review and test this big PR.

**Comment by dorkamotorka** [comment]
I tested with web routes for 5 hours. No error.

**Comment by JanezCim** [comment]
Tested it multiple times on a real robot in a small room and it seems to be working alright. I need to test in TP still in a large area still and with different speeds. Im testing with this launch, can you confirm all parameters are ok?

```
<?xml version="1.0"?>
<!-- This launch file is the collection of the move_basic params relevant for Qualcomm robot -->

<launch>
  <node name="move_basic" pkg="move_basic" type="move_basic"  output="screen">    
    <param name="base_frame" value="base_link"/>
    <param name="reverse_without_turning_threshold" value="0.1"/>
    <param name="angular_tolerance" value="0.08"/>
    <param name="linear_tolerance" value="0.1"/>
    <param name="velocity_threshold" value="0.2"/>
    <param name="max_linear_velocity" value="0.3"/>
    <param name="max_turning_velocity" value="0.5"/>
    <param name="lateral_kp" value="0.3"/>
    <param name="lateral_kd" value="0.5"/>
    <param name="linear_gain" value="0.5"/>
    <param name="rotational_gain" value="3.5"/>
    <param name="linear_acceleration" value="0.4"/>
    <param name="abort_timeout" value="0.5"/>    

    <param name="forward_obstacle_threshold" value="0.2"/>
    <param name="robot_back_length" value="0.33"/>
    <param name="min_side_dist" value="0.2"/>
    <param name="robot_width" value="0.2"/>
    <param name="robot_front_length" value="0.2"/>

    <remap from="/move_base/goal" to="move_base/goal"/>
    <remap from="/move_base_simple/goal" to="move_base_simple/goal"/>
    <remap from="/plan" to="plan"/>
    <remap from="/obstacle_viz" to="obstacle_viz"/>
    <remap from="/cmd_vel" to="cmd_vel"/>
    <remap from="/lateral_error" to="lateral_error"/>
    <remap from="/obstacle_distance" to="obstacle_distance"/>
    <remap from="/plan" to="plan"/>
    <remap from="/scan" to="scan_filtered"/>
    
  </node>
</launch>

```

**Comment by dorkamotorka** [comment]
image
The parameter I tested with are on the image. In my opinion the parameters should be tuned a bit more, since it still happens on some goal that the robot triggers abort: Moving away from goal. So thread with cautious. Nevertheless, if anyone finds suitable parameters write the down and publish it here or something :)

**Review by JanezCim — APPROVED** [review]
Tested on a real robot works well enough :)

**Review by MoffKalast — APPROVED** [review]
Tested in gazebo as well, doesn't seem like there are any issues in general.

**Comment by JanezCim** [comment]
I guess it would be of great value if original author of move_basic would take a detailed look at this before we merge right? @rohbotics @jim2v

**Comment by jim-v** [comment]
> I guess it would be of great value if original author of move_basic would take a detailed look at this before we merge right? @rohbotics @jim2v

I could take a look, but won't be able to get to it for a few days.

**Comment by dorkamotorka** [comment]
This would be great if Jim and Rohan can take a look. We can hold it off, no problem.

File: src/move_basic.cpp
Comment (jim-v) [inline src/move_basic.cpp:643]: It's interesting this was done in terms of time.  Does it work well?

**Review by jim-v — APPROVED** [review]
The changes look reasonable to me.  It's nice to see some of the special case code has been cleaned up.

File: src/move_basic.cpp
Comment (dorkamotorka) [reply src/move_basic.cpp:643]: I added a timeout, since localization with LiDAR is jumping +-0.02 and when we get really close to goal, the robot might thought it is over the goal(cos(angleRemaining) becomes negative since the goal is behind - at least what the localization think it is if it flaky)). If there where no timeout, goal would be immediately aborted as soon as the localization would jump "over the goal". This happened often when testing(of course when we where trying to reach the goal better than the localization accuracy).

Comment (MoffKalast) [reply src/move_basic.cpp:643]: I think that's an argument towards retrying a goal instead of just aborting it, so that the goal is reached in any case and merely triggers a recalculation when the angle doesn't seem to be sane.

**Comment by rohbotics** [comment]
I don't have time to properly review the code, but I like the overall changes and there are already approvals so you can go ahead and merge @tp4348

**Pull request merged**
